odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,242 @@
+
+        try:
+            _logger.info('Calling "%s" on "%s"', job['method'], job['res_model'])
+            records = env[job['res_model']].browse(job['res_ids'])
+            result = getattr(records, job['method'])(*job['args'], **job['kwargs'])
+            json.dumps(result)  # ensure result is serializable
+            records.flush()
+        except Exception as exc:
+            job['state'] = 'failed'
+            job['payload'] = ir_async._handle_exception(exc)
+        else:
+            _logger.debug("Done, result is %s", result)
+            if result is None:
+                job['state'] = 'done'
+            else:
+                job['state'] = 'succeeded'
+                job['payload'] = ir_async._json_response(result)
+
+        ir_async._post_process(job)
+
+    def _pre_process(self, job):
+        pass
+
+    def _post_process(self, job):
+        pass
+
+    def _merge_traceback(self, tb):
+        """ Merge the given traceback with all previous level of execution """
+        old_tb = self.env.context.get('async_traceback', '').splitlines(keepends=True)
+        if old_tb:
+            def fnc_process_job_not_reached(line):
+                return not re.search(f'in {self._process_job.__name__}\\W', line)
+            old_tb.append("Traceback of async job (most recent call last):\n")
+            old_tb.extend(itertools.dropwhile(fnc_process_job_not_reached, tb))
+            tb = old_tb
+        return "".join(tb)
+
+    def _handle_exception(self, exception):
+        header, *tb = traceback.format_exc().splitlines(keepends=True)
+        exc_tb = header + self._merge_traceback(tb)
+
+        if isinstance(exception, odoo.exceptions.UserError):
+            _logger.warning(exception)
+        else:
+            _logger.error("Failed to process async job\n%s", exc_tb)
+
+        error = {
+            'code': 200,
+            'message': "Odoo Server Error",
+            'data': serialize_exception(exception),
+        }
+        error['data']['debug'] = exc_tb
+        return self._json_response(error=error)
+
+    def _json_response(self, result=None, error=None):
+        response = {}
+        if result is not None:
+            response['result'] = result
+        if error is not None:
+            response['error'] = error
+        return response
+
+    @api.autovacuum
+    def _vacuum_terminated_tasks(self):
+        self._cr.execute("""
+            SELECT id
+            FROM ir_async
+            WHERE state in ('failed', 'done')
+               OR (state = 'succeeded'
+                   AND write_date + interval '3 day' < LOCALTIMESTAMP)
+            FOR UPDATE SKIP LOCKED
+            """)
+        ids = tuple(self._cr.fetchall())
+        if ids:
+            self._cr.execute("""
+                DELETE FROM ir_async
+                WHERE id in %s
+                """, (ids,))
+        _logger.info("Vacuumed %d terminated asynchronous tasks", len(ids))
+
+    def complete(self):
```

Comment:
This method does nothing if the user_id is not the right one.  Shouldn't it fail instead?